ts/state/ducks/proBackendData.ts

```diff
@@ -0,0 +1,186 @@
+import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit';
+import type { StateType } from '../reducer';
+import {
+  ProProofResultType,
+  ProRevocationsResultType,
+  ProStatusResultType,
+} from '../../session/apis/pro_backend_api/types';
+import ProBackendAPI from '../../session/apis/pro_backend_api/ProBackendAPI';
+
+type RequestState<D = unknown> = {
+  isFetching: boolean;
+  // Shortcut for `isFetching && !data`
+  isLoading: boolean;
+  // Shortcut for `!!error`
+  isError: boolean;
+  error: string | null;
+  t: number;
+  data: D | null;
+};
+
+const defaultRequestState = {
+  isFetching: false,
+  isLoading: false,
+  isError: false,
+  error: null,
+  t: 0,
+  data: null,
+} satisfies RequestState;
+
+export type ProBackendDataState = {
+  proof: RequestState<ProProofResultType>;
+  revocations: RequestState<ProRevocationsResultType>;
+  proStatus: RequestState<ProStatusResultType>;
+};
+
+export const initialProBackendDataState: ProBackendDataState = {
+  proof: defaultRequestState,
+  revocations: defaultRequestState,
+  proStatus: defaultRequestState,
+};
+
+const fetchProProofFromProBackend = createAsyncThunk(
+  'proBackendData/fetchProProof',
+  async (_, payloadCreator): Promise<RequestState<ProProofResultType>> => {
+    try {
+      if (window.sessionFeatureFlags?.debugServerRequests) {
+        window.log.info(
+          `[networkData/fetchInfoFromSeshServer] starting ${new Date().toISOString()}`
+        );
+      }
+
+      const state = payloadCreator.getState() as StateType;
+      if (state.proBackendData.proof.isFetching) {
+        throw new Error('already fetching');
+      }
+
+      payloadCreator.dispatch(
+        proBackendDataSlice.actions.setIsFetching({ key: 'proof', result: true })
+      );
+      payloadCreator.dispatch(
+        proBackendDataSlice.actions.setIsLoading({
+          key: 'proof',
+          result: !state.proBackendData.proof.data,
+        })
+      );
+
+      const response = await ProBackendAPI.getProProof();
+      if (!response) {
+        throw new Error('Data fetch failed');
+      }
+
+      if (response.status_code !== 200) {
+        return {
+          data: state.proBackendData.proof.data,
+          error: 'TODO: get the error',
+          isError: true,
+          isFetching: false,
+          isLoading: false,
+          t: response.t,
+        };
+      }
+
+      return {
+        data: response.result,
+        error: null,
+        isError: false,
+        isFetching: false,
+        isLoading: false,
+        t: response.t,
+      };
+    } finally {
+      payloadCreator.dispatch(
+        proBackendDataSlice.actions.setIsFetching({ key: 'proof', result: false })
+      );
+      payloadCreator.dispatch(
+        proBackendDataSlice.actions.setIsLoading({ key: 'proof', result: false })
+      );
```

Comment:
this will be called when you throw an error to prevent multiple runs in parallel and will not be correct.

Reply:
on line53 you do
```ts
if (state.proBackendData.proof.isFetching) {
        throw new Error('already fetching');
      }
```
This will make the call go into the `finally`, and mark the call as not fetching, which is incorrect.


What I mean is that if you do, through the UI or another way
```ts
dispatch(fetchProProofFromProBackend())
dispatch(fetchProProofFromProBackend())
```

the first one will mark `state.proBackendData.proof.isFetching` as true, and the second one will go right away in the finally and mark all of the calls as not fetching

Reply:
on line53 you do
```ts
if (state.proBackendData.proof.isFetching) {
        throw new Error('already fetching');
      }
```
This will make the call go into the `finally`, and mark the call as not fetching, which is incorrect.


What I mean is that if you do, through the UI or another way
```ts
dispatch(fetchProProofFromProBackend())
dispatch(fetchProProofFromProBackend())
```

the first one will mark `state.proBackendData.proof.isFetching` as true, and the second one will go right away in the finally and mark all of the calls as not fetching

Reply:
I think this should work yeah